### ISSUE TITLE
parameters

### ISSUE THREAD

**Issue by chase-moskal** [malthe/ts-postgres#24]
from the readme,

> Query parameters use the format $1, $2 etc.

do the parameters have to be numbered? i'd much prefer to use names, as i'm familiar with using parameter placeholders like `SELECT * FROM whatever WHERE :lol = 5` or something like that, right?

i'm just really hoping i don't have to worry about how things are ordered, i'm an options-object kind of guy

**Comment by malthe** [comment]
Yes, it is probably possible to have a preprocessor that scans through the statement, replacing `:<param-name>` with a position-based expression and then map it correctly on query execute.

But if you look at https://github.com/brianc/node-postgres/issues/268, there are plenty of reasons not to want something like that – opting instead for doing this on a higher level; for example, a query builder could provide such an interface.

**Comment by chase-moskal** [comment]
oh. well that makes sense. thanks for sharing that link!

maybe we could whip up some sugar for named params?

would you rather that be in a separate repo/package, or would you be interested for some kind of `sugar/` directory here for experimenting with these kinds of extras?

**Comment by chase-moskal** [comment]
> preprocessor that scans through the statement

i just read that, and thought about tagged template literals at the same time — and got a nerd-boner!

**Comment by malthe** [comment]
It might be possible to implement this sugar using a tokenizer only and not have to actually _parse_ the query.

**Comment by chase-moskal** [comment]
i'm thinking something along these lines ought be possible!

```ts
const result = await query`SELECT * FROM table WHERE column = ${value}`
```

how do you fancy?

**Comment by malthe** [comment]
That looks good. People understand the escaping of those literals because of the tagged literal syntax.

And it's far better than `:param = ...` which would be hard to implement correctly and is as foreign to JS/TypeScript as it is to PostgreSQL.

**Comment by chase-moskal** [comment]
i'll look into putting together a pull request for you, to add a `sugar` directory, so we can experiment and mull it over :)
